File: spaced_rep_june_2022/coinChange.js
var coinChange = function(coins, amt) {
    let dp = [...new Array(amt+1)].fill(Infinity)
    dp[0] = 0
    for (let i = 1; i <= amt; i++) {
        for(let c of coins) {
            if (i - c >= 0 && i - c <= i) {
                dp[i] = Math.min(dp[i], dp[i-c] + 1)
            }
        }
    }
    if (dp[amt] === Infinity) return -1
    return dp[amt]
};

File: java_problems_2024/coin_change.java
package java_problems_2024;
import java.util.Arrays;
class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = 10^4 +1;
        int[] dp = new int[amount+1];
        Arrays.fill(dp, max);
        dp[0] = 0;

        for (int c : coins) {
            for (int a = 1; a <= amount; a++) {
                if (a -c >= 0) {
                    dp[a] = Math.min(dp[a], dp[a -c] + 1);
                }
            }
        }
        return dp[amount] == max ? -1 : dp[amount];
    }
}
